pkg/debugger/debugger.go

```diff
@@ -0,0 +1,211 @@
+			return list, fmt.Errorf("error sending message over WebSocket: %w", err)
+		}
+
+		eventAction := <-wsw.eventSignal
+		actionType, _ := eventAction["type"].(string)
+		value, _ := eventAction["value"].(string)
+		acceptedEvents, _ = EventAction(actionType, value, acceptedEvents, event)
+	}
+	return acceptedEvents, nil
+}
+
+// EventAction decides, based on the input what exactly is done next with the current event
+func EventAction(
+	actionType string,
+	_ string,
+	acceptedEvents *events.EventList,
+	currentEvent *eventpb.Event,
+) (*events.EventList, error) {
+	if actionType == "accept" {
+		acceptedEvents.PushBack(currentEvent)
+	}
+	return acceptedEvents, nil
+}
+
+func (wsw *WSWriter) HandleClientSignal(signal map[string]interface{}) {
+	wsw.eventSignal <- signal
+}
+
+// newWSWriter creates a new WSWriter that establishes a websocket connection
+func newWSWriter(port string) *WSWriter {
+	fmt.Println("Starting newWSWriter")
+
+	// Create a new WSWriter object
+	wsWriter := &WSWriter{
+		upgrader: websocket.Upgrader{
+			ReadBufferSize:  ReadBufferSize,
+			WriteBufferSize: WriteBufferSize,
+		},
+		eventSignal: make(chan map[string]interface{}),
+	}
+
+	// Create an Async go routine that waits for the connection
+	go func() {
+		http.HandleFunc("/ws", func(w http.ResponseWriter, r *http.Request) {
+			wsWriter.upgrader.CheckOrigin = func(r *http.Request) bool { return true } // Allow opening the connection by HTML file
+			conn, err := wsWriter.upgrader.Upgrade(w, r, nil)
+			if err != nil {
+				panic(err)
+			}
+
+			fmt.Println("WebSocket connection established")
+
+			// Update the attribute of the WSWriter object with the established connection
+			wsWriter.conn = conn
+			// go routine for incoming messages
+			go func() {
+				defer func(conn *websocket.Conn) {
+					err := conn.Close()
+					if err != nil {
+						panic(err)
+					}
+				}(conn) // Ensure the connection is closed when the function exits
+
+				for {
+					_, message, err := conn.ReadMessage()
+					if err != nil {
+						break
+					}
+
+					var signal map[string]interface{}
+					err = json.Unmarshal(message, &signal)
+					if err != nil {
+						continue
+					}
+					wsWriter.HandleClientSignal(signal)
+				}
+			}()
+		})
+
+		server := &http.Server{
+			Addr:         port,
+			Handler:      nil,
+			ReadTimeout:  5 * time.Second,
+			WriteTimeout: 10 * time.Second,
+			IdleTimeout:  15 * time.Second,
+		}
+
+		err := server.ListenAndServe()
+		if err != nil {
+			panic(err)
+		}
+		select {}
+	}()
+	return wsWriter
+}
+
+// writerFactory creates and returns a WebSocket-based event writer
+// It determines the port for the WebSocket server based on the given nodeID and a base port value
+func writerFactory(_ string, nodeID t.NodeID, _ logging.Logger) (eventlog.EventWriter, error) {
+	ownPort, err := strconv.Atoi(string(nodeID))
+	if err != nil {
+		panic(err)
+	}
+	ownPort += BasePort
```

Comment:
It would be a bit cleaner to ignore the node ID here and pass the port number to the writer in a different way. For example by defining this function in-line when creating the interceptor, and using a closure to pass the port number directly.